akka-http-core/src/main/scala/akka/http/impl/engine/http2/client/PersistentConnection.scala

```diff
@@ -31,8 +36,11 @@ private[http2] object PersistentConnection {
    *  * generate error responses with 502 status code
    *  * custom attribute contains internal error information
    */
-  def managedConnection(connectionFlow: Flow[HttpRequest, HttpResponse, Any]): Flow[HttpRequest, HttpResponse, NotUsed] =
-    Flow.fromGraph(new Stage(connectionFlow))
+  def managedConnection(connectionFlow: Flow[HttpRequest, HttpResponse, Future[OutgoingConnection]], maxAttempts: Int): Flow[HttpRequest, HttpResponse, NotUsed] =
```

Comment:
What would it be used for?
The returned `Flow` would fail in the scenario where the materialized `Future` would, so the only benefit would be to signal that it was once successfully connected (which could probably be observed from the `connectionFlow`)

Reply:
Right now, if you put this `Flow` after some other akka streams elements, and the stream fails, you don't know which (if any) of the requests have started processing. If materialization would produce a future that fails in that case at least you would be sure none of the requests were processed.

Indeed you'll still have to handle the stream failing for the scenario's when the connection fails after running successfully for a while. Hard to say how important the special case of not connecting at all is.

Anyway, not needed for this PR, but something to think about before we unmark this API 'ApiMayChange'.

Reply:
:+1: I think the most important part is that we emit some descriptive error at all when we fail after max attempts.

Reply:
I.e. I think it would be somewhat weird to treat the first connection attempt differently than others later.